users/miles2go/babblePaste.h

```diff
@@ -0,0 +1,357 @@
+/*  A library to output the right key shortcut in any common app. 
+Given a global variable babble_mode to show the environment and a 
+key that calls the paste macro, do the right type of paste. 
+
+Setting the bable_mode is done by another macro, or TBD interaction with the host. 
+
+Huge thanks to https://en.wikipedia.org/wiki/Table_of_keyboard_shortcuts
+and jeebak & algernon's keymap
+*/
+
+#pragma once
+#include "quantum.h"
+
+
+
+#ifdef USE_BABBLEPASTE
+
+
+void set_babble_mode( uint8_t id);
+void babble_mode_increment(void);
+void babble_mode_decrement(void);
+void babble_led_user(void);
+
+// manually re-order these if you want to set the order or default. 
+enum babble_modes{ 
+#ifdef BABL_MAC
+	BABL_MAC_MODE,
+#endif 
+#ifdef BABL_READMUX
+	BABL_READMUX_MODE,
+#endif 
+#ifdef BABL_WINDOWS
+	BABL_WINDOWS_MODE,
+#endif
+#ifdef BABL_VI
+	BABL_VI_MODE,
+#endif 
+#ifdef BABL_LINUX
+	BABL_LINUX_MODE,
+#endif 
+#ifdef BABL_EMACS
+	BABL_EMACS_MODE,
+#endif
+#ifdef BABL_CHROMEOS
+	BABL_CHROMEOS_MODE,
+#endif
+BABL_MODEMAX 
+} ;
+
+//void babble_led_user( uint8_t id) 
+
+///Hacks to make it easier to create sendstring macros
+
+//"outer" versions wrap text
+#define OMCTL(arg) SS_DOWN(X_LCTRL) arg SS_UP(X_LCTRL)
```

Comment:
I believe these can be simplified with [combos](https://github.com/qmk/qmk_firmware/blob/57540af102d034396c9f41a5a6e69ead8b10ba99/docs/feature_combo.md): `SS_DOWN(X_FOO) arg SS_UP(X_FOO)` to `tap_code16(FOO(arg))`

Reply:
That won't work if `arg` is a string literal.

Reply:
Hmm. I think you might need to use the [stringizing operator](https://docs.microsoft.com/en-us/cpp/preprocessor/stringizing-operator-hash?view=vs-2019) for a string literal, so `tap_code16(FOO(#arg))`

Reply:
I'd be concerned that the stringizing would break some combination of inner wrapper & wrapped strings. This way is not elegant but it's clear.

Reply:
I just remembered part of why I did the #define macros this way - the result is a string that can be stored in flash with PROGMEM.  On smaller devices like the 32u4, space starts to matter. Using tap_code16 would mix functions & text and change the storage type.

Reply:
@xplusplus stringizing still won't work; `tap_code()` and friends operate on USB HID keycode values, not ASCII characters.